packages/contracts/contracts/Unirep.sol

```diff
@@ -99,47 +99,40 @@ contract Unirep is IUnirep, VerifySignature {
     }
 
     /**
-     * @dev User signs up by provding a zk proof outputting identity commitment and new gst leaf.
-     * msg.sender must be attester
-     */
-    function userSignUp(uint256[] memory publicSignals, uint256[8] memory proof)
-        public
-    {
-        uint256 attesterId = publicSignals[2];
-        // only allow attester to sign up users
-        if (uint256(uint160(msg.sender)) != attesterId)
-            revert AttesterIdNotMatch(uint160(msg.sender));
-        // Verify the proof
-        if (!signupVerifier.verifyProof(publicSignals, proof))
-            revert InvalidProof();
-
-        uint256 identityCommitment = publicSignals[0];
+     * Use this if your application has custom signup proof logic.
+     * e.g. to insert a non-zero data field in the state tree leaf
+     **/
+    function manualUserSignUp(
```

Comment:
I think it can be public input in `signup.circom`
```js
    signal input data[FIELD_COUNT];
    ....
    for (var x = 0; x < FIELD_COUNT; x++) {
      leaf_hasher.data[x] <== data[x];
    }
```
it can make sure all `stateTreeLeaf` are not a random hash

or we can simply not use `signup` circuit
because the `attesterId`, `epoch` are verified on-chain 
and `leaf` and the `idCommitment` is not verified (like `manualUserSignUp`)

Reply:
I'm confused, what is the suggestion here?

I thought about making it a public input to the `signup` circuit, but applications might want to calculate the data in a specialized way. Like an application might want to give the user secret data by doing `H(idSecret, data)`.

We do need to figure out how to get the starting data into the `UserState`. Right now if a user manually signs up with a custom state tree leaf the pre-image isn't recorded anywhere.

Reply:
Right. but what is the purpose of `signup` proof?
attesters can use `manualUserSignUp` to save gas.
and make sure the users `stateTreeLeaf = hash(data)`

Reply:
Ah that's interesting. I think the biggest case where we want the signup proof is smart contract attesters. Those attesters _could_ implement the signup verifier in their own contract logic, but I think it's nice to include it in the unirep core contract as a base signup case.

I think having both options is nice, but I agree that we should use the manual case more often.

Reply:
Maybe in the `manualSignUp` instead of `stateTreeLeaf` we accept `uint[]` that are the state tree pre-image and calculate the state tree leaf in the contract?

Reply:
I think it also work in `signup.circom`
The attester smart contract can check `H(idSecret, data)` with their own verifier
and the attester submits the proof with `H(idSecret, data)` is one of the public signals
then the `userSignUp` can verify the `signup` proof
(so no `manualSignUp`)

Reply:
Would this involve verifying 2 proofs onchain?

Reply:
either two proofs or like `H(leaf, data)` (public data) which can be solved by a hash on-chain.